### DISCUSSION TITLE
How to execute commands in .chezmoi.toml.tmpl with params from chezmoi vars

### DISCUSSION THREAD

**Discussion by dgudim** [twpayne/chezmoi#3191] [Q&A] [answered]
What I am trying to do is automatically infer gpg key from user's email, I have tried several things, none of them worked. This is the config I currently have which does not work

```
{{- $git_username := promptStringOnce . "git.git_username" "Git username" -}}
{{- $git_email := promptStringOnce . "git.git_email" "Git email" -}}

[scriptEnv]
    GIT_EMAIL = {{ $git_email | quote }}

[data.git]
    git_username = {{ $git_username | quote }}
    git_email = {{ $git_email | quote }}

[data.gpg]
    signingkey = {{ output "sh" "-c" "gpg --with-colons --locate-keys $GIT_EMAIL | grep pub | cut -d : -f 5" | quote }}

{{ if eq .gpg.signingkey "" }}
{{  writeToStdout "Warning, signing key for " }}
{{  writeToStdout .git.git_email }}
{{  writeToStdout " wasn't found\n" }}
{{ else }}
{{  writeToStdout "Inferred signing key: " }}
{{  writeToStdout .gpg.signingkey }}
{{  writeToStdout "\n" }}
{{ end }}
```

**Comment by twpayne** [comment]
> I have tried several things, none of them worked

What exactly have you tried and what exactly happened? You need to be more specific if you want help. People cannot read minds.

I suspect the problem is that you're mixing up template data (every variable beginning with a `.`) and template variables (every variable beginning with a `$`). Specifically, the template data (e.g. `.gpg.signingkey`) is fixed while the template is executed, whereas you're hoping to set it in the `[data.gpg]` section. You probably need to assign the signing key to a `$` variable like `$gpg_signingkey` and then use this instead of `.gpg.signingkey`.

**Reply by dgudim** [reply]
> What exactly have you tried and what exactly happened? 

Sorry for that, I have tried the above method, setting the variable as you suggested and playing with `output` command parameters. In all the cases it either complained about missing/undefined variables or the key was empty

> You probably need to assign the signing key to a $ variable like $gpg_signingkey

Yes, I have also tried that, namely this
This always produces "\\"\\"" instead of signing key

I suspect it has something to do with `scriptEnv` not setting the variable or whatever or I am just dumb

```
[scriptEnv]
    GIT_EMAIL = {{ $git_email | quote }}

{{$signingkey := output "sh" "-c" "gpg --with-colons --locate-keys $GIT_EMAIL | grep pub | cut -d : -f 5" | quote}}

[data.gpg]
    signingkey = {{ $signingkey | quote }}
```

**Reply by twpayne** [reply]
You're still not giving enough information. What commands do I need to run to reproduce the problem?

**Reply by dgudim** [reply]
Paste this:
```
{{- $git_email := promptStringOnce . "git.git_email" "Git email" -}}

[scriptEnv]
    GIT_EMAIL = {{ $git_email | quote }}

{{$signingkey := output "sh" "-c" "gpg --with-colons --locate-keys $GIT_EMAIL | grep pub | cut -d : -f 5"}}

{{ if eq $signingkey "" }}
{{  writeToStdout "Warning, signing key for " }}
{{  writeToStdout $git_email }}
{{  writeToStdout " wasn't found\n" }}
{{ else }}
{{  writeToStdout "Inferred signing key: " }}
{{  writeToStdout $signingkey }}
{{  writeToStdout "\n" }}
{{ end }}
```
into `.chezmoi.toml.tmpl` in the chezmoi's root and run `chezmoi init`, input some valid email which has a linked key in your gpg keyring, observe the output, the inferred key will be empty

**Reply by twpayne** [reply, marked as answer]
I have not run the commands, but:

    {{- $git_email := promptStringOnce . "git.git_email" "Git email" -}}

This sets the `$git_email` local template variable.

    [scriptEnv]
        GIT_EMAIL = {{ $git_email | quote }}

This sets the `$GIT_EMAIL` environment variable to `$git_email` for **the next time that the script is run**.

    {{$signingkey := output "sh" "-c" "gpg --with-colons --locate-keys $GIT_EMAIL | grep pub | cut -d : -f 5"}}

This uses the current value of the `$GIT_EMAIL` environment variable. If you're running `chezmoi init` for the first time then this environment variable will **not be set**.

Fundamentally, I think the confusion here is between imperative code and declarative data. For example:

    [scriptEnv]
        GIT_EMAIL = {{ $git_email | quote }}

Does **not** set the `$GIT_EMAIL` environment variable to the contents of the `$git_email` template variable now. It instead adds text to the generated chezmoi config file that means that the **next** time that chezmoi runs and reads its config file the `$GIT_EMAIL` environment variable will be set for scripts.

**Reply by dgudim** [reply]
Sorry for wasting your time. I ended up doing this, the `output` command is quite limiting

```
{{ output "sh" "-c" "echo 'echo $(gpg --with-colons --locate-keys $1 | grep pub | cut -d : -f 5)' > /tmp/infer.sh"}}
{{$signingkey := output "bash" "/tmp/infer.sh" $git_email | trim }}
```

**Reply by 2-4601** [reply]
How about this:
```
{{ $GIT_EMAIL := printf "%s=%s" "GIT_EMAIL" $git_email }}
{{ $signingkey := output "env" $GIT_EMAIL "sh" "-c" "gpg --with-colons --locate-keys $GIT_EMAIL | grep pub | cut -d : -f 5" }}
```

**Reply by dgudim** [reply]
Thanks! I didn't think about that. This also works
